client/modules/cm-tarnation/src/parser/parser.ts

```diff
@@ -0,0 +1,257 @@
+   * @param pending - The chunks to parse.
+   * @param editorContext - The CodeMirror editor parse context to use.
+   */
+  constructor(
+    language: TarnationLanguage,
+    context: ParserContext,
+    input: Input,
+    region: ParseRegion,
+    editorContext?: EditorParseContext,
+    pending: Chunk[] = []
+  ) {
+    this.language = language
+    this.context = context
+    this.region = region
+    this.pending = pending
+    this.embeddedHandler = new EmbeddedHandler(language, context, input, editorContext)
+  }
+
+  get pos() {
+    return this.pending?.[this.context.index]?.pos ?? this.region.from
+  }
+
+  get parserDone() {
+    // doesn't work right now :L
+    // return this.context.index >= this.pending.length || this.pos >= this.region.to
+    return this.context.index >= this.pending.length
+  }
+
+  /** True if the parser has finished parsing already. */
+  get done() {
+    return this.parserDone && this.embeddedHandler.done
+  }
+
+  /** Executes a parse step. */
+  private parse() {
+    const ctx = this.context
+    const chunk = this.pending[ctx.index]
+
+    // we want to cache the context before we process the chunk
+    // this is the _starting_ context, not the ending context
+    this.language.cache.attach(this.context.serialize(), chunk)
+
+    const tokens = chunk.compile()
+    for (let idx = 0; idx < tokens.length; idx++) {
+      // bit of a hack, but for some reason TS isn't typw narrowing the switch below
+      // so we'll just act like this is a MappedToken, and then assert as any for
+      // EmbedToken
+      const t = tokens[idx] as MappedToken
+
+      switch (typeof t[0]) {
+        // embed token
+        case "string": {
+          const token: LezerToken = [0, t[1], t[2], -1, Tree.empty]
+          ctx.buffer.add(token)
+          ctx.stack.increment()
+          this.embeddedHandler.add(token, t[0])
+          break
+        }
+        // mapped token (default)
+        default: {
+          /*
+           * this upcoming code entirely avoids iterator methods, like destructuring
+           * thus, it's entirely unreadable
+           * I've left comments describing what a destructured approach looks like,
+           * but without actually using it.
+           * doing it this way has a decent speed boost but yeah it looks awful
+           */
+
+          // const [type, from, to, open, close] = token
+
+          // add open nodes to stack
+          // this doesn't affect the buffer at all, but now we can watch for
+          // when another node closes one of the open nodes we added
+          // if (open) {
+          if (t[3]) {
+            // for (let i = 0; i < open.length; i++) {
+            for (let i = 0; i < t[3].length; i++) {
+              // const [id, inclusive] = open[i]
+              const o = t[3][i]
+              // ctx.stack.push(id, inclusive ? from : to, type ? (inclusive ? 0 : -1) : 0)
+              ctx.stack.push(o[0], o[1] ? t[1] : t[2], t[0] ? (o[1] ? 0 : -1) : 0)
```

Comment:
Might be more readable if it was split onto multiple lines:
```suggestion
              ctx.stack.push(
                o[0],
                o[1] ? t[1] : t[2],
                t[0] ? (o[1] ? 0 : -1) : 0),
              )
```

Reply:
good idea, made a commit where I did the same to the commented out destructured version